rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs

```diff
@@ -0,0 +1,732 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::client::auth::{
+    AuthSchemeId, HttpAuthScheme, SharedAuthOptionResolver, SharedHttpAuthScheme,
+};
+use crate::client::connectors::SharedConnector;
+use crate::client::identity::{ConfiguredIdentityResolver, SharedIdentityResolver};
+use crate::client::interceptors::SharedInterceptor;
+use crate::client::orchestrator::SharedEndpointResolver;
+use crate::client::retries::{RetryClassifiers, SharedRetryStrategy};
+use aws_smithy_async::rt::sleep::SharedAsyncSleep;
+use aws_smithy_async::time::SharedTimeSource;
+use std::fmt;
+
+pub(crate) static EMPTY_RUNTIME_COMPONENTS_BUILDER: RuntimeComponentsBuilder =
+    RuntimeComponentsBuilder::new("empty");
+
+macro_rules! merge {
+    (Option $other:ident . $name:ident => $self:ident) => {
+        $self.$name = $other.$name.clone().or($self.$name.take());
+    };
+    (Vec $other:ident . $name:ident => $self:ident) => {
+        if !$other.$name.is_empty() {
+            $self.$name.extend($other.$name.iter().cloned());
+        }
+    };
+}
+macro_rules! builder_field_value {
+    (Option $self:ident . $name:ident) => {
+        $self.$name
+    };
+    (Option $self:ident . $name:ident required) => {
+        $self.$name.ok_or(BuildError(concat!(
+            "the `",
+            stringify!($name),
+            "` runtime component is required"
+        )))?
+    };
+    (Vec $self:ident . $name:ident) => {
+        $self.$name
+    };
+    (Vec $self:ident . $name:ident atLeastOneRequired) => {{
+        if $self.$name.is_empty() {
+            return Err(BuildError(concat!(
+                "at least one `",
+                stringify!($name),
+                "` runtime component is required"
+            )));
+        }
+        $self.$name
+    }};
+}
+macro_rules! rc_type {
```

Comment:
maybe write this out? I thought this had something to do with rc

Reply:
Addressed in 68e834d622a4a77afb388f8cc6cef81cc299cc4a